### PR TITLE
Implemented string-distance, string-equal, compare-strings

### PR DESCRIPTION
**Pull request by Ki11erRabbit**
I implemented the above mentioned functions. I plan on doing the remaining after I get feedback on my style. I also had to implement a few methods on some types to allow for getting their byte data for `string-equal`.
I also commented out an implementation I found in `search.rs` for `string-equal`. I will remove it after I get feedback as well as adjusting `CONTRIBUTING.org` to reflect the implemented functions.

Feel free to leave any comments.

### CODE REVIEW

File: src/fns.rs [hunk 1]
```diff
@@ -615,6 +615,188 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
     }
 }
 
+#[defun]
+pub(crate) fn string_equal<'ob>(s1: Object<'ob>, s2: Object<'ob>) -> Result<Option<bool>> {
+    let s1 = match s1.untag() {
+        ObjectType::String(x) => x.as_bytes(),
+        ObjectType::ByteString(x) => x.inner(),
+        ObjectType::Symbol(x) => x.get().as_bytes(),
+        _ => bail!("Wrong type argument: stringp, {s1}"),
```
Comment (CeleritasCelery) [inline src/fns.rs:624]: We have a `ArgErr` type that can be used to capture this type.

File: src/fns.rs [hunk 2]
```diff
@@ -615,6 +615,188 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
...
+
+    if s1.len() != s2.len() {
+	    return Ok(None);
+    }
+    if s1 == s2 {
+	    Ok(Some(true))
+    } else {
+	    Ok(None)
+    }
+}
```
Comment (CeleritasCelery) [inline src/fns.rs]: We can just return `Ok(s1 == s2)`

File: src/fns.rs [hunk 3]
```diff
@@ -615,6 +615,188 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
...
+
+#[defun]
+pub(crate) fn string_distance<'ob>(
+    string1: Object<'ob>,
+    string2: Object<'ob>,
+    bytecompare: Option<Object<'ob>>
+) -> Result<Object<'ob>> {
+    let bytecompare = if let Some(x) = bytecompare {
+        !matches!(x.untag(), ObjectType::NIL)
```
Comment (CeleritasCelery) [inline src/fns.rs:733]: We can just use `x != NIL`

File: src/fns.rs [hunk 4]
```diff
@@ -615,6 +615,188 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
...
+    };
+
+    Ok(distance.into())
+}
+
+#[inline]
+pub(crate) fn levenshtein_distance<T: PartialEq>(
+    s1: &mut dyn Iterator<Item = T>,
+    s2: &mut dyn Iterator<Item = T>
+) -> i64 {
```
Comment (CeleritasCelery) [inline src/fns.rs]: Is there a reason this has be dynamically dispatched instead of generic?

File: src/fns.rs [hunk 5]
```diff
@@ -615,6 +615,188 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
     }
 }
 
+#[defun]
+pub(crate) fn string_equal<'ob>(s1: Object<'ob>, s2: Object<'ob>) -> Result<Option<bool>> {
```
Comment (CeleritasCelery) [inline src/fns.rs:619]: Why does the return type need to be `Option<bool>` instead of just bool?

File: src/fns.rs [hunk 6]
```diff
@@ -615,6 +615,187 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
...
+    if s1 == s2 {
+        Ok(Some(true))
+    } else {
+        Ok(None)
+    }
+}
+
+#[defun]
+pub(crate) fn compare_strings<'ob>(
+    string1: Object<'ob>,
```
Comment (CeleritasCelery) [inline src/fns.rs:645]: We should be able to use the more specific type of `string1: &str` here to make the code cleaner

File: src/fns.rs [hunk 7]
```diff
@@ -615,6 +615,187 @@ pub(crate) fn elt<'ob>(sequence: Object<'ob>, n: usize, cx: &'ob Context) -> Res
...
+        }
+        leading += 1;
+    }
+
+    Ok(true.into())
+}
+
+#[defun]
+pub(crate) fn string_distance<'ob>(
+    string1: Object<'ob>,
```
Comment (CeleritasCelery) [inline src/fns.rs:728]: can just use `&str`.

**Comment by Ki11erRabbit** [comment]
I should have fixed the issues you have brought up.

**Comment by CeleritasCelery** [comment]
Thanks! This looks good

**Pull request merged**